op-node/rollup/interop/managed/system.go

```diff
@@ -378,60 +378,131 @@ func (m *ManagedMode) Reset(ctx context.Context, lUnsafe, xUnsafe, lSafe, xSafe,
+
+	x := int(latestUnsafe.Number - target)
+	if x < 0 {
+		logger.Warn("Latest unsafe block is older than target, using latest unsafe for search")
+		target = latestUnsafe.Number
+	} else if x != 0 {
+		// search space:
+		// ------------------------------------------------------------------------------------
+		// target.Number |  offset=0   offset=1   offset=2  ...  offset = x-1 = latestUnsafe   |  offset=x  (doesn't exist)
+		// false         |  t/f        t/f        t/f       ...  t/f                           |  true
+		// ------------------------------------------------------------------------------------
+		offset, err := search(x, func(i int) (bool, error) {
+			return isInvalidBlock(target + 1 + uint64(i))
+		})
+		if err != nil {
+			return eth.L2BlockRef{}, err
+		}
+		// offset == 0 => search returns 0 if f(i) returned true for all i in [0, x) => all blocks we checked are invalid
+		// offset == x => search returns x if f(i) returned false for all i in [0, x) => all blocks we checked are valid
+
+		if offset != 0 {
+			logger.Info("Found last valid block with binary search", "validNum", target+uint64(offset))
+
+			valid, err := m.l2.L2BlockRefByNumber(ctx, target+uint64(offset))
+			if err != nil {
+				return eth.L2BlockRef{}, err
+			}
+
+			logger.Info("Last valid L2 block ref", "valid", valid)
+			return valid, nil
+		}
 	}
 
-	// Start from latestUnsafe and go down to find the latest valid block, since L1 reorgs aren't usually that deep.
-	for n := latestUnsafe.Number; ; n-- {
-		if n == l2UnsafeTarget.Number-1 {
+	// offset == 0 or x == 0, so go from target backwards indefinitely until we find a valid block
+	// since L1 reorgs aren't usually that deep.
+	for n := target; ; n-- {
+		if n == target-1 {
 			logger.Warn("No valid unsafe block found up to target, searching further")
 		}
 		// Exit loop if context is done - other errors are only logged and loop continues.
 		if cerr := ctx.Err(); cerr != nil {
 			return eth.L2BlockRef{}, fmt.Errorf("context done: %w", cerr)
 		}
 
-		current, err := m.l2.L2BlockRefByNumber(ctx, n)
+		isInvalid, err := isInvalidBlock(n)
 		if err != nil {
-			logger.Warn("Failed to get L2 block ref, trying previous", "err", err, "current", n)
-			continue
+			return eth.L2BlockRef{}, err
 		}
 
-		logger := logger.New("current", current, "currentL1Origin", current.L1Origin)
-		// make sure L1 origin hasn't been reorged
-		l1Origin, err := m.l1.L1BlockRefByNumber(ctx, current.L1Origin.Number)
-		if err != nil {
-			logger.Warn("Failed to get origin L1 block ref, trying previous", "err", err)
-			continue
-		}
-		if l1Origin.Hash != current.L1Origin.Hash {
-			logger.Warn("L1 block was reorged on this block, trying previous", "newL1Origin", l1Origin)
-			continue
+		if !isInvalid {
+			logger.Info("Found last valid block", "validNum", n)
+
+			valid, err := m.l2.L2BlockRefByNumber(ctx, n)
+			if err != nil {
+				return eth.L2BlockRef{}, err
```

Comment:
Behavior is modified here too, as we now return on error (also within the binary search lambda).